contracts/src/Teleporter/upgrades/TeleporterRegistry.sol

```diff
@@ -37,6 +111,85 @@ contract TeleporterRegistry is WarpProtocolRegistry {
+     * Requirements:
+     *
+     * - `version` must be a valid version, i.e. greater than 0 and not greater than the latest version.
+     */
+    function getTeleporterFromVersion(
+        uint256 version
+    ) external view returns (ITeleporterMessenger) {
+        return ITeleporterMessenger(getAddressFromVersion(version));
+    }
+
+    /**
+     * @dev Gets the address of a protocol version.
+     * Requirements:
+     *
+     * - `version` must be a valid version, i.e. greater than 0 and not greater than the latest version.
+     */
+    function getAddressFromVersion(
+        uint256 version
+    ) public view returns (address) {
+        require(version != 0, "TeleporterRegistry: zero version");
+        require(
+            version <= latestVersion,
+            "TeleporterRegistry: invalid version"
+        );
+        return _versionToAddress[version];
+    }
+
+    /**
+     * @dev Gets the version of the given `protocolAddress`.
+     * If `protocolAddress` is not a registered protocol address, returns 0, which is an invalid version.
+     */
+    function getVersionFromAddress(
+        address protocolAddress
+    ) public view returns (uint256) {
+        require(
+            protocolAddress != address(0),
+            "TeleporterRegistry: zero protocol address"
+        );
+        return _addressToVersion[protocolAddress];
+    }
+
+    /**
+     * @dev Adds the new protocol version address to the registry.
+     * Updates latest version if the version is greater than the current latest version.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
+     * Note: `protocolAddress` doesn't have to be a contract address, this is primarily
+     * to support the case in which we want to register a new protocol address meant for a security patch
+     * before the contract is deployed, to prevent the vulnerabilitiy from being exposed before the registry update.
+     * Requirements:
+     *
+     * - `version` is not zero
+     * - `version` is not already registered
+     * - `protocolAddress` is not zero address
+     */
+    function _addToRegistry(ProtocolRegistryEntry memory entry) private {
+        require(entry.version != 0, "TeleporterRegistry: zero version");
+        // Check that the version has not previously been registered.
+        require(
+            _versionToAddress[entry.version] == address(0),
+            "TeleporterRegistry: version already exists"
+        );
+        require(
+            entry.protocolAddress != address(0),
+            "TeleporterRegistry: zero protocol address"
+        );
+
+        _versionToAddress[entry.version] = entry.protocolAddress;
+        _addressToVersion[entry.protocolAddress] = entry.version;
+        emit AddProtocolVersion(entry.version, entry.protocolAddress);
+
+        // Set latest version if the version is greater than the current latest version.
+        if (entry.version > latestVersion) {
+            latestVersion = entry.version;
```

Comment:
Should we have a `LatestVersionUpdated` event?

Reply:
added